Display cause stacks in tooltips.

mstange opened the pull request:
Fixes #64.

<img width="1552" alt="screen shot 2017-11-27 at 1 10 42 pm" src="https://user-images.githubusercontent.com/961291/33281824-8f3ddc66-d374-11e7-8b1a-e86f285841ef.png">

This is ready for review but not completely ready for landing.

Remaining work:

 * [x] Change flow type of the `getTooltipContents` to something more specific than `Function`
 * [ ] Try to reduce the duplication of the `getTooltipContents` implementations
 * [x] Add a snapshot test
 * [x] Should probably move the `GeckoSyncProfile` type definition to the gecko-profile type file.
 * [ ] Implement printing the stack to the console when a marker is clicked (could also be a follow-up)

julienw commented on src/profile-logic/process-profile.js:
```diff
@@ -533,6 +534,22 @@ function _processMarkers(geckoMarkers: GeckoMarkerStruct): MarkersTable {
+            const newData = Object.assign({}, m);
+            delete newData.stack;
+            if ('stack' in m && m.stack && m.stack.samples.data.length > 0) {
+              const syncProfile = m.stack;
```
what's this name `syncProfile`? Why not `stack` ?

mstange replied:
Renamed to stack.

julienw commented on src/profile-logic/old-cleopatra-profile-format.js:
```diff
@@ -183,6 +186,29 @@ function _convertThread(
     const marker = thread.markers[i];
     const markerIndex = markers.length++;
+    const data = marker.data;
+    if (data && 'stack' in data) {
```
so we used to have the stack in cleopatra already ?

mstange replied:
Yes we did! Only for styles markers, though. I'm not sure when the reflow cause was added. The old cleopatra exposed the cause callstack in two ways: With a regular tooltip on the markers in the header, and in the markers list where the "syncProfile" call tree was displayed as a subtree of the marker node.

julienw commented on src/components/marker-chart/index.js:
```diff
@@ -2,14 +2,17 @@
 import { updateProfileSelection } from '../../actions/profile-view';
+import { getImplementationFilter } from '../../reducers/url-state';
 
+import type { Element as ReactElement } from 'react';
```
nit: you can just use `React.Element` as we use `import * as React from 'react'` above.

julienw commented on src/components/marker-chart/Canvas.js:
```diff
@@ -6,10 +6,10 @@
 import TextMeasurement from '../../utils/text-measurement';
 import { BLUE_40 } from '../../utils/colors';
 
+import type { Element as ReactElement } from 'react';
```
nit: you can just use React.Element as we use import * as React from 'react' above.

julienw commented on src/components/marker-chart/Canvas.js:
```diff
@@ -43,6 +43,7 @@ type Props = {
   markerTimingRows: MarkerTimingRows,
   rowHeight: CssPixels,
   markers: TracingMarker[],
+  getTooltipContents: TracingMarker => ReactElement<any>,
```
maybe use `React.Node` as return value (this means: anything than can be rendered by react).
`React.Element<any>` is part of `React.Node` so Flow will happily accept the passed function.

mstange replied:
Then I also need to adjust `getHoveredMarkerInfo`, but the comment above it explicitly says that that's ok, so I'm going to do that :)

julienw commented on src/components/shared/MarkerTooltipContents.js:
```diff
@@ -12,10 +12,13 @@ import {
   formatBytes,
   formatValueTotal,
 } from '../../utils/format-numbers';
+import Backtrace from '../header/Backtrace';
```
maybe `Backtrace` should be in `shared` ? I think that nothing in `shared` should depend on anything else than `shared` (except maybe `utils`); moreover it's not used in the header only, but also in the marker view.

julienw commented on src/components/header/Backtrace.js:
```diff
@@ -0,0 +1,74 @@
+      thread,
+      implementationFilter,
+      callNodePathForStack(thread, cause.stack)
+    );
```
sometimes it's empty here when we use the JS-only implementation filter; should we display something in that case? Maybe we shouldn't display the `at` string in that case too, so maybe move it here ? Or move this call to the parent component ? I don't know :)
This would get better once we have pseudoframes.

Also are we getting the proper stack for everything done by JS? I'm thinking for example of when we switch a class, then a restyle is performed. In that case, I we'd see a JS stack for the 'styles' marker, and another one for the 'reflow' marker?

julienw commented on src/components/header/Backtrace.css:
```diff
@@ -0,0 +1,25 @@
+  margin: 5px;
+  padding: 0;
+  list-style: none;
+  display: block;
```
Wow, I never knew `<ol>` was an inline element !

Can you please move this line at the top of the rule, so that it's easier to read ?

mstange replied:
It's `display:block` by default, what I did here was unnecessary. I'll remove it. Thanks for catching it.

julienw commented on src/components/header/Backtrace.css:
```diff
@@ -0,0 +1,25 @@
+  margin: 0;
+  padding: 0;
+  list-style: none;
+  display: block;
```
same :)

mstange replied:
This one actually has a use: `li` is `display: list-item` by default. The `list-style: none` declaration makes `display: list-item` act very similarly (or maybe even identically?) to `display: block`, so strictly speaking, you'd only need either `display: block` or `list-style: none`, but I always specify both out of habit.

julienw commented on src/components/header/Backtrace.css:
```diff
@@ -0,0 +1,25 @@
+  list-style: none;
+  display: block;
+  max-width: 30em;
+  text-overflow: ellipsis;
```
not for this patch, but we need a way to have the ellipsis at the middle. For example, extract the hostname and display it, and then display the end of the path only. We don't really need the scheme here, I think.

mstange replied:
Yeah, that's good to keep in mind.

julienw commented on src/components/header/Backtrace.js:
```diff
@@ -0,0 +1,74 @@
+
+// @flow
+
+import React, { PureComponent } from 'react';
```
please use the new convention of `import * as React from 'react'` and use `React.PureComponent` below.

mstange replied:
Ok, sure.

julienw commented on src/components/header/Backtrace.js:
```diff
@@ -0,0 +1,74 @@
+  implementationFilter: ImplementationFilter,
+};
+
+type State = {};
```
just don't specify a State at all, and it will be `void` (Flow's `undefined`).

Here you say it can be any object, this is not what you want.

julienw commented on src/components/header/Backtrace.js:
```diff
@@ -0,0 +1,74 @@
+  return path;
+}
+
+class Backtrace extends PureComponent<Props, State> {
```
As said above, please leave out the State parameter.

julienw replied:
or instead maybe you can make this a functional component instead of a class component ?

mstange replied:
I've made it a functional component.

julienw commented on src/components/header/Backtrace.js:
```diff
@@ -0,0 +1,74 @@
+  thread: Thread,
+  cause: CauseBacktrace,
+  implementationFilter: ImplementationFilter,
+};
```
nit: please define props like this:
```javascript
type Props = {|
  +thread: Thread,
  ...
|};
```
This makes the props stricter (we can't provide more property by mistake) and read-only. Newer Flow (v0.59) has a $ReadOnly utility type for the same thing, but we can use `+` now so let's use it :) I try to use this style in newer components for now but I'd like to convert them all eventually.

mstange commented:
I've addressed most comments.

You're right that the tooltip is usually not large enough to display complete C++ stack traces.

Once we have better pseudostacks, I'd like to display an "abridged stack" that starts with a lot of detail  at the top and gets progressively less detailed towards the bottom. For example:

 - The first two or three C++ frames
 - [...] (some indication that we're leaving stuff out)
 - The next pseudo frame (e.g. "Element.setAttribute")
 - Two or three JS frames
 - [...]
 - The rest of the JS stack grouped by host
 - [...]
 - The bottom-most frame that called into JS
 - Maybe the name of the runnable that the event loop is running at the moment

gregtatum commented:
Can you drop in a screenshot to the PR next time you're working on this?

mstange commented:
I've added a screenshot to the first comment of this issue. I think this is ready to land.

julienw commented on src/components/header/ProfileThreadJankOverview.js:
```diff
@@ -4,24 +4,43 @@
+        implementationFilter={implementationFilter}
+      />
+    );
+  };
```
about the "HOC" I had in mind (actually not a HOC), I think it was something similar to the function you already have, except make it a functional stateless component:
```javascript
const TooltipContents = ({ item }: { string }) =>
  <MarkerTooltipContents
     marker={item}
     threadName={threadName}
     thread={thread}
     implementationFilter={implementationFilter}
  />;
```

So that in `IntervalMarkerOverview` you can use it like this:
```
render() {
  const { TooltipContents, ... } = this.props;
  return (
    stuff
    <TooltipContents item={XXX}/>
  );   
}
```
So this is very similar to what you have but it's maybe more "react". So I think this is better to do it like this :)

Yet It provides no performance benefit and I think the best idea is to memoize this. We already use `memoize-immutable` in this project so you can just use it here, no additional dependency.

julienw replied:
To use `memoize` you'll need to have a separate function that will generate this stateless component.

This separate function should take the `thread`, the `threadName` and the `implementationFilter`. (I mention it because I considered taking `state` and `threadIndex` but then the memoization wouldn't work when `state` changes). Then maybe you could even extract it and reuse it in both places already memoized after all :)

(note: memoize caches all calls by default, which I think is fine for us)

julienw replied:
at this point, maybe it's good to just make MarkerTooltipContents a connected component :)

mstange commented on src/components/shared/Tooltip.css:
```diff
@@ -64,3 +64,14 @@
+
+.tooltipDetailsBackTrace {
+
+}
```
oops

gregtatum commented:
This is getting pretty stale. Should we close it, or should we have someone finish up the changes here (@julienw I'd be fine if you fixed your review points and merged yourself)

julienw commented:
I think we should finish this up, as this exposes interesting data. Realistically I won't be able to come to it before the end of this week, but we can include it in our next sprint to finish it, unless Markus wants to finish it.

gregtatum commented:
Works for me, let's leave it open for now.

julienw commented:
rebased in https://github.com/devtools-html/perf.html/pull/756, I'll close this :)

Now I'll work on fixing the comments

julienw commented on src/types/markers.js:
```diff
@@ -16,36 +18,32 @@ export type GPUMarkerPayload = {
+export type PaintProfilerMarkerTracing_Gecko = {
   type: 'tracing',
-  startTime: Milliseconds, // Same as cpustart
-  endTime: Milliseconds, // Same as cpuend
```
ping @mstange about this; if these comments are still accurate I can add them back but I'd like to be sure.

mstange replied:
Yes, that was on purpose. "cpustart" and "cpuend" are fields that don't exist on ProfilerMarkerTracing objects; they only existed on GPUMarkerPayload objects but even those have been removed now. (We should remove the typing for them from this file as well, but probably in a separate PR.)

julienw commented on src/types/markers.js:
```diff
@@ -16,36 +18,32 @@ export type GPUMarkerPayload = {
-    | 'ForwardTransaction'
-    | 'NotifyDidPaint'
-    | 'LayerTransaction'
-    | 'Composite',
```
ping @mstange about this: was this wrong before ?

mstange commented on src/types/gecko-profile.js:
```diff
@@ -27,6 +27,17 @@ export type GeckoMarkerStruct = {
+  pid: number,
+  markers: GeckoMarkers,
+  samples: GeckoSamples,
+};
```
Yes, this is the v9 Gecko format, which is still going to be at v9 after this patch. So it's typing the current format.

This patch is only changing the version of the "processed profile" format. It's leaving the version of the "Gecko profile" format unchanged.

Before: v9 Gecko, v9 processed
After: v9 Gecko, v10 processed

julienw replied:
OK, got it now, thanks !
